### PR TITLE
chatbot-Instruction Adherence

### PR DESCRIPTION
**Pull request by mansiaimon**

### CODE REVIEW

File: examples/chatbot.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+from aimon.client import Client
+from llama_index.embeddings.openai import OpenAIEmbedding
+from dotenv import load_dotenv
+import requests
+import json
+from aimon import Config
+import logging
+import time
+
+load_dotenv()
```
Comment (pjoshi30) [inline examples/chatbot.py:22]: Lets remove the need to get this from .env and get the required credentials from the UI.

File: examples/chatbot.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+email = os.getenv('EMAIL')
```
Comment (pjoshi30) [inline examples/chatbot.py:24]: for now this is required. lets get this from the UI as well. Once the new SDK is launched, we wont need this.

File: examples/chatbot.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
+logging.basicConfig(level=logging.DEBUG)
+
+nest_asyncio.apply()
+os.environ['SSL_CERT_FILE'] = certifi.where()
+
+# Download UBER dataset
```
Comment (pjoshi30) [inline examples/chatbot.py:32]: can you move line 32 to line 51 inside an init function? Additionally, also create a class called `Chatbot` that has all of this logic inside it.

File: examples/chatbot.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+for year in years:
+    storage_context = StorageContext.from_defaults()
+    cur_index = VectorStoreIndex.from_documents(
+        doc_set[year],
+        storage_context=storage_context,
+    )
+    index_set[year] = cur_index
+    storage_context.persist(persist_dir=f"./storage/{year}")
+
+individual_query_engine_tools = [
```
Comment (pjoshi30) [inline examples/chatbot.py:69]: line 69 to line 92 can be refactored in to a `construct_agent()` function

File: examples/chatbot.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+    openai.api_key = openai_api_key
+    contexts = []
+    relevance_scores = []
+    input_text = f"Instructions: {instructions}\nQuery: {user_query}"
+
+    logging.info("Sending request to OpenAI API...")
+    chat_response = agent.chat(input_text)
+    logging.info("Received response from OpenAI API.")
+
+    if hasattr(chat_response, 'source_nodes'):
```
Comment (pjoshi30) [inline examples/chatbot.py:118]: lets refactor line 118 through 129 into a function. Maybe call it `get_source_docs`

File: examples/chatbot_app.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+            chatbot_response = "Error occurred while processing your request."
+
+        end_time = time.time()
+        elapsed_time = end_time - start_time
+        logging.info(f"Time taken for response: {elapsed_time:.2f} seconds")    
+
+        if chatbot_response == "Error occurred while processing your request.":
+            st.write(chatbot_response)
+        else:
+            st.session_state['hallucination_score'] = hallucination_score
```
Comment (pjoshi30) [inline examples/chatbot_app.py:82]: where are these objects (hallucination_score, toxicity etc.) that are stored within `st.session_state` used?

**Comment by pjoshi30** [comment]
Added some comments on the code. Apart from that, here is a few things we can do to make this a bit better:

1. Attach the results of the .detect() call to each conversation flow so that one can refer back to the score for a conversation if they'd like to do so.
2. Investigate methods to reduce the latency app's from the current baseline of 30+ seconds. We should know what is taking the most time and some ideas to improve it.
3. Get the API keys from the UI instead of two places: `.env` file and the UI.
4. Compute the embeddings of the Uber dataset and store them on local disk unless they change the embedding model. This will ensure that for anyone who uses this chatbot, they will not repeatedly compute the embeddings every time they start and stop the app.

**Comment by mansiaimon** [comment]
> st.session_state['hallucination_score'] = hallucination_score

point 1 and 4 have been done

File: examples/chatbot.py (hunk 2 above)
Comment (mansiaimon) [reply examples/chatbot.py:29]: done

File: examples/chatbot_app.py (hunk 6 above)
Comment (mansiaimon) [reply examples/chatbot_app.py]: these objects aren't used anymore

File: examples/chatbot.py [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+    else:
+        combined_context = "\n".join(top_contexts)
+
+    data_to_send = validate_and_format_json([{
+        "context": combined_context,
+        "generated_text": chat_response.response,
+        "instructions": "\n".join(instructions)
+    }])
+
+    max_retries = 3
```
[the remark this answers is not in the document]
Comment (mansiaimon) [reply examples/chatbot.py:195]: done

File: examples/chatbot.py (hunk 5 above)
Comment (mansiaimon) [reply examples/chatbot.py:167]: done

File: examples/chatbot.py (hunk 4 above)
Comment (mansiaimon) [reply examples/chatbot.py:116]: done

File: examples/chatbot.py (hunk 3 above)
Comment (mansiaimon) [reply examples/chatbot.py:51]: done

Comment (mansiaimon) [reply examples/chatbot.py:51]: created class EmbeddingsManager that has all the logic inside it

**Pull request closed without merging**